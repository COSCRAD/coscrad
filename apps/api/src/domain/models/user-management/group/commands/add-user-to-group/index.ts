export * from './add-user-to-group.command';
export * from './add-user-to-group.command-handler';
