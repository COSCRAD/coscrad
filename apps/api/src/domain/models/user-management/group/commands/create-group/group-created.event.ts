import { BaseEvent } from '../../../../shared/events/base-event.entity';

export class GroupCreated extends BaseEvent {
    type = 'USER_GROUP_CREATED';
}
