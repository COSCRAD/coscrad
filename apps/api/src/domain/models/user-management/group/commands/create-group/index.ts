export * from './create-group.command';
export * from './create-group.command-handler';
