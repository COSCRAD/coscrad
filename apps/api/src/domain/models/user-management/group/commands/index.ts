export * from './add-user-to-group';
export * from './create-group';
