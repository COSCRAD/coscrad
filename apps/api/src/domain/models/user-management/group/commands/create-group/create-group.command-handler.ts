import { CommandHandler } from '@coscrad/commands';
import { Inject } from '@nestjs/common';
import { InternalError } from '../../../../../../lib/errors/InternalError';
import { REPOSITORY_PROVIDER_TOKEN } from '../../../../../../persistence/constants/persistenceConstants';
import { DTO } from '../../../../../../types/DTO';
import { ResultOrError } from '../../../../../../types/ResultOrError';
import { Valid } from '../../../../../domainModelValidators/Valid';
import buildInstanceFactory from '../../../../../factories/utilities/buildInstanceFactory';
import { IIdManager } from '../../../../../interfaces/id-manager.interface';
import { IRepositoryForAggregate } from '../../../../../repositories/interfaces/repository-for-aggregate.interface';
import { IRepositoryProvider } from '../../../../../repositories/interfaces/repository-provider.interface';
import { IUserRepository } from '../../../../../repositories/interfaces/user-repository.interface';
import { AggregateId } from '../../../../../types/AggregateId';
import { AggregateType } from '../../../../../types/AggregateType';
import { InMemorySnapshot } from '../../../../../types/ResourceType';
import buildInMemorySnapshot from '../../../../../utilities/buildInMemorySnapshot';
import { BaseCreateCommandHandler } from '../../../../shared/command-handlers/base-create-command-handler';
import { BaseEvent } from '../../../../shared/events/base-event.entity';
import { validAggregateOrThrow } from '../../../../shared/functional';
import { CoscradUserGroup } from '../../entities/coscrad-user-group.entity';
import { CreateGroup } from './create-group.command';
import { GroupCreated } from './group-created.event';

@CommandHandler(CreateGroup)
export class CreateGroupCommandHandler extends BaseCreateCommandHandler<CoscradUserGroup> {
    readonly aggregateType = AggregateType.userGroup;

    protected readonly repositoryForCommandsTargetAggregate: IRepositoryForAggregate<CoscradUserGroup>;

    protected readonly userRepository: IUserRepository;

    constructor(
        @Inject(REPOSITORY_PROVIDER_TOKEN)
        protected readonly repositoryProvider: IRepositoryProvider,
        @Inject('ID_MANAGER') protected readonly idManager: IIdManager
    ) {
        super(repositoryProvider, idManager);

        this.repositoryForCommandsTargetAggregate = repositoryProvider.getUserGroupRepository();

        this.userRepository = repositoryProvider.getUserRepository();
    }

    protected createNewInstance({
        aggregateCompositeIdentifier: { id },
        label,
        description,
    }: CreateGroup): ResultOrError<CoscradUserGroup> {
        const createDto: DTO<CoscradUserGroup> = {
            type: AggregateType.userGroup,
            id,
            label,
            description,
            // You must run 'ADD_USER_TO_GROUP' to add users
            userIds: [],
        };

        return buildInstanceFactory(CoscradUserGroup)(createDto);
    }

    protected buildEvent(
        command: CreateGroup,
        eventId: string,
        systemUserId: AggregateId
    ): BaseEvent {
        return new GroupCreated(command, eventId, systemUserId);
    }

    protected async fetchRequiredExternalState(): Promise<InMemorySnapshot> {
        const [existingUserGroupSearchResult, allUserSearchResult] = await Promise.all([
            this.repositoryForCommandsTargetAggregate.fetchMany(),
            this.userRepository.fetchMany(),
        ]);

        const existingGroups = existingUserGroupSearchResult.filter(validAggregateOrThrow);

        const users = allUserSearchResult.filter(validAggregateOrThrow);

        return buildInMemorySnapshot({
            user: users,
            userGroup: existingGroups,
        });
    }

    protected validateExternalState(
        state: InMemorySnapshot,
        instance: CoscradUserGroup
    ): InternalError | Valid {
        return instance.validateExternalState(state);
    }
}
