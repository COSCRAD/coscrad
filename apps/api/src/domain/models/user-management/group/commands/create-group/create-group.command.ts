import { Command } from '@coscrad/commands';
import { NestedDataType, NonEmptyString, UUID } from '@coscrad/data-types';
import { Equals } from 'class-validator';
import { AggregateCompositeIdentifier } from '../../../../../types/AggregateCompositeIdentifier';
import { AggregateType } from '../../../../../types/AggregateType';
import { ICreateCommand } from '../../../../shared/command-handlers/interfaces/create-command.interface';

class GroupCompositeId {
    /**
     * This is a bit of a hack. It circumvents our `CoscradDataTypes` and may
     * cause problems for
     * - Schema management
     * - Anyone using our API directly (not via front-end)
     *
     * The simple answer is that you always have to tack on an
     * `aggregateCompositeIdentifier`.
     */
    @Equals(AggregateType.userGroup)
    type = AggregateType.userGroup;

    @UUID({
        label: 'ID',
        description: 'unique identifier',
    })
    id: string;
}

@Command({
    type: 'CREATE_USER_GROUP',
    label: 'Create User Group',
    description: 'Creates a new user group',
})
export class CreateGroup implements ICreateCommand {
    @NestedDataType(GroupCompositeId, {
        label: 'Composite Identifier',
        description: 'system-wide unique identifier',
    })
    readonly aggregateCompositeIdentifier: AggregateCompositeIdentifier<
        typeof AggregateType.userGroup
    >;

    @NonEmptyString({
        label: 'label',
        description: 'the name of the user group',
    })
    readonly label: string;

    // TODO Consider sharing the annotation with the domain model
    @NonEmptyString({
        label: 'description',
        description: 'a summary of the purpose of this group (what its members have in common)',
    })
    readonly description: string;
}
