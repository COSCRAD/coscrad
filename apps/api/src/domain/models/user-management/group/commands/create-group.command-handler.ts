import { CommandHandler } from '@coscrad/commands';
import { Inject } from '@nestjs/common';
import { InternalError } from '../../../../../lib/errors/InternalError';
import { RepositoryProvider } from '../../../../../persistence/repositories/repository.provider';
import { DTO } from '../../../../../types/DTO';
import { ResultOrError } from '../../../../../types/ResultOrError';
import { Valid } from '../../../../domainModelValidators/Valid';
import buildInstanceFactory from '../../../../factories/utilities/buildInstanceFactory';
import { IIdManager } from '../../../../interfaces/id-manager.interface';
import { IRepositoryForAggregate } from '../../../../repositories/interfaces/repository-for-aggregate.interface';
import { IUserRepository } from '../../../../repositories/interfaces/user-repository.interface';
import { AggregateType } from '../../../../types/AggregateType';
import { InMemorySnapshot } from '../../../../types/ResourceType';
import buildInMemorySnapshot from '../../../../utilities/buildInMemorySnapshot';
import { BaseCreateCommandHandler } from '../../../shared/command-handlers/base-create-command-handler';
import { BaseEvent } from '../../../shared/events/base-event.entity';
import { validAggregateOrThrow } from '../../../shared/functional';
import { CoscradUserGroup } from '../entities/coscrad-user-group.entity';
import { CreateGroup } from './create-group.command';
import { GroupCreated } from './group-created.event';

@CommandHandler(CreateGroup)
export class CreateGroupCommandHandler extends BaseCreateCommandHandler<CoscradUserGroup> {
    readonly aggregateType = AggregateType.userGroup;

    // we may want to rename this `repositoryForCommandTargetAggregate`
    protected readonly repository: IRepositoryForAggregate<CoscradUserGroup>;

    protected readonly userRepository: IUserRepository;

    constructor(
        protected readonly repositoryProvider: RepositoryProvider,
        @Inject('ID_MANAGER') protected readonly idManager: IIdManager
    ) {
        super(repositoryProvider, idManager);

        this.repository = repositoryProvider.getUserGroupRepository();

        this.userRepository = repositoryProvider.getUserRepository();
    }

    protected createNewInstance({
        id,
        label,
        description,
    }: CreateGroup): ResultOrError<CoscradUserGroup> {
        const createDto: DTO<CoscradUserGroup> = {
            type: AggregateType.userGroup,
            id,
            label,
            description,
            // You must run 'ADD_USER_TO_GROUP' to add users
            userIds: [],
        };

        return buildInstanceFactory(CoscradUserGroup)(createDto);
    }

    protected eventFactory(command: CreateGroup, eventId: string): BaseEvent {
        return new GroupCreated(command, eventId);
    }

    protected async fetchRequiredExternalState(): Promise<InMemorySnapshot> {
        const [existingUserGroupSearchResult, allUserSearchResult] = await Promise.all([
            this.repository.fetchMany(),
            this.userRepository.fetchMany(),
        ]);

        const existingGroups = existingUserGroupSearchResult.filter(validAggregateOrThrow);

        const users = allUserSearchResult.filter(validAggregateOrThrow);

        return buildInMemorySnapshot({
            users,
            userGroups: existingGroups,
        });
    }

    protected validateExternalState(
        state: InMemorySnapshot,
        instance: CoscradUserGroup
    ): InternalError | Valid {
        return instance.validateExternalState(state);
    }
}
