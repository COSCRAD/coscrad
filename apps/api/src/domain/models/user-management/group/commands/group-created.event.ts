import { AggregateId } from '../../../../types/AggregateId';
import { BaseEvent } from '../../../shared/events/base-event.entity';
import { CreateGroup } from './create-group.command';

export class GroupCreated extends BaseEvent {
    type = 'USER_GROUP_CREATED';

    constructor(command: CreateGroup, eventId: AggregateId) {
        super(command, eventId);
    }
}
