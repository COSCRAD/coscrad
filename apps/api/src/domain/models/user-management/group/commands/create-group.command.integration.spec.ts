import { CommandHandlerService, FluxStandardAction } from '@coscrad/commands';
import setUpIntegrationTest from '../../../../../app/controllers/__tests__/setUpIntegrationTest';
import { NotAvailable } from '../../../../../lib/types/not-available';
import { ArangoConnectionProvider } from '../../../../../persistence/database/arango-connection.provider';
import generateRandomTestDatabaseName from '../../../../../persistence/repositories/__tests__/generateRandomTestDatabaseName';
import TestRepositoryProvider from '../../../../../persistence/repositories/__tests__/TestRepositoryProvider';
import { DTO } from '../../../../../types/DTO';
import { IIdManager } from '../../../../interfaces/id-manager.interface';
import { AggregateId } from '../../../../types/AggregateId';
import buildEmptyInMemorySnapshot from '../../../../utilities/buildEmptyInMemorySnapshot';
import { assertCreateCommandSuccess } from '../../../__tests__/command-helpers/assert-create-command-success';
import { CommandAssertionDependencies } from '../../../__tests__/command-helpers/types/CommandAssertionDependencies';
import { CoscradUserGroup } from '../entities/coscrad-user-group.entity';
import { CreateGroup } from './create-group.command';
import { CreateGroupCommandHandler } from './create-group.command-handler';

const commandType = 'CREATE_USER_GROUP';

const buildValidCommandFSA = (id: AggregateId): FluxStandardAction<DTO<CreateGroup>> => ({
    type: commandType,

    payload: {
        id,
        label: 'teachers',
        description: 'this group is for teachers from community schools',
    },
});

const initialState = buildEmptyInMemorySnapshot();

describe('CreateGroup', () => {
    let testRepositoryProvider: TestRepositoryProvider;

    let commandHandlerService: CommandHandlerService;

    let arangoConnectionProvider: ArangoConnectionProvider;

    let idManager: IIdManager;

    let commandAssertionDependencies: CommandAssertionDependencies;

    beforeAll(async () => {
        ({ testRepositoryProvider, commandHandlerService, idManager, arangoConnectionProvider } =
            await setUpIntegrationTest({
                ARANGO_DB_NAME: generateRandomTestDatabaseName(),
            }));

        commandHandlerService.registerHandler(
            commandType,
            new CreateGroupCommandHandler(testRepositoryProvider, idManager)
        );

        commandAssertionDependencies = {
            testRepositoryProvider,
            idManager,
            commandHandlerService,
        };
    });

    afterAll(async () => {
        await arangoConnectionProvider.dropDatabaseIfExists();
    });

    beforeEach(async () => {
        await testRepositoryProvider.testSetup();
    });

    afterEach(async () => {
        await testRepositoryProvider.testTeardown();
    });
    describe('when the command is valid', () => {
        it('should succeed', async () => {
            await assertCreateCommandSuccess(commandAssertionDependencies, {
                buildValidCommandFSA,
                initialState,
                checkStateOnSuccess: async ({ id }: CreateGroup) => {
                    const userGroupSearchResult = await testRepositoryProvider
                        .getUserGroupRepository()
                        .fetchById(id);

                    expect(userGroupSearchResult).toBeInstanceOf(CoscradUserGroup);

                    // This allows us to visually  inspect the payload => new instance mapping
                    expect(userGroupSearchResult).toMatchSnapshot();

                    // Ensure the id has been marked as used
                    const idStatus = await idManager.status(id);

                    expect(idStatus).toBe(NotAvailable);
                },
            });
        });
    });
});
