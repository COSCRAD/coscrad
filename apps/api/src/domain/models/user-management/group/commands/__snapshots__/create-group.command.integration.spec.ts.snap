// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateGroup when the command is valid should succeed 1`] = `
CoscradUserGroup {
  "description": "this group is for teachers from community schools",
  "eventHistory": Array [
    Object {
      "meta": Object {
        "dateCreated": 1586044800000,
        "id": "41fb2d7f-c483-4e09-a1f0-e9909a6b0002",
      },
      "payload": Object {
        "description": "this group is for teachers from community schools",
        "id": "41fb2d7f-c483-4e09-a1f0-e9909a6b0001",
        "label": "teachers",
      },
      "type": "USER_GROUP_CREATED",
    },
  ],
  "getCompositeIdentifier": [Function],
  "id": "41fb2d7f-c483-4e09-a1f0-e9909a6b0001",
  "label": "teachers",
  "type": "userGroup",
  "userIds": Array [],
}
`;
